File: app/src/main/java/net/nekobako/twidere/extension/aniuta/NowPlaying.java
package net.nekobako.twidere.extension.aniuta;

import android.content.*;
import android.os.Bundle;
import org.mariotaku.twidere.constant.IntentConstants;


public class NowPlaying {
	private Listener listener;


	public void setListener(Listener listener) {
		this.listener = listener;
	}

	public void fetchMusicMetadata(Context context) {
		context.sendOrderedBroadcast(new Intent(MusicPlayerListenerService.ACTION_FETCH_MUSIC_METADATA), null, new BroadcastReceiver() {
			@Override
			public void onReceive(Context context, Intent intent) {
				if(NowPlaying.this.listener != null) {
					Bundle extras = this.getResultExtras(false);
					NowPlaying.this.listener.onMusicMetadataFetched(extras != null ? (MusicMetadata)extras.getParcelable("music_metadata") : null);
				}
			}
		}, null, 0, null, null);
	}

	public Intent createTweetIntent(Context context, MusicMetadata musicMetadata) {
		Intent intent = new Intent();
		Bundle extras = new Bundle();
		extras.putString(IntentConstants.EXTRA_APPEND_TEXT, context.getResources().getString(R.string.tweet, musicMetadata.getTitle(), musicMetadata.getArtist()));
		extras.putParcelable(IntentConstants.EXTRA_IMAGE_URI, musicMetadata.getArtworkUri(context));
		intent.putExtras(extras);
		return intent;
	}


	public interface Listener {
		void onMusicMetadataFetched(MusicMetadata musicMetadata);
	}
}

File: app/src/main/java/net/nekobako/twidere/extension/aniuta/NowPlaying.kt
package net.nekobako.twidere.extension.aniuta

import android.content.*
import android.os.Bundle
import org.mariotaku.twidere.constant.IntentConstants


class NowPlaying {
	companion object {
		private val TWIDERE_PACKAGE_NAME = "org.mariotaku.twidere"
	}


	fun fetchMusicMetadata(context: Context, callback: (MusicMetadata?) -> Unit) {
		context.sendOrderedBroadcast(Intent(MusicPlayerListenerService.ACTION_FETCH_MUSIC_METADATA), null, object : BroadcastReceiver() {
			override fun onReceive(context: Context, intent: Intent?) {
				callback.invoke(this.getResultExtras(false)?.getParcelable("music_metadata"))
			}
		}, null, 0, null, null)
	}

	fun createTweetIntent(context: Context, musicMetadata: MusicMetadata): Intent {
		val intent = Intent()
		val extras = Bundle()
		val uri = musicMetadata.getArtworkUri(context)
		extras.putString(IntentConstants.EXTRA_APPEND_TEXT, context.resources.getString(R.string.tweet, musicMetadata.title, musicMetadata.artist))
		extras.putParcelable(IntentConstants.EXTRA_IMAGE_URI, uri)
		intent.putExtras(extras)
		context.grantUriPermission(TWIDERE_PACKAGE_NAME, uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
		return intent
	}
}
